### ISSUE TITLE
Need a range data output from the backend database attached to the root range item

### ISSUE THREAD

**Issue by skybristol** [usgs-bcb/gap_species#22]
Drop the file here:

https://www.sciencebase.gov/catalog/item/5951527de4b062508e3b1e79

**Comment by skybristol** [comment]
Contact information and citations for the ranges

**Comment by nmtarr** [comment]
I put together a draft csv file with metadata information on the range data for ScienceBase.  I modeled it after the one for the habitat maps.  I put the [csv file](https://github.com/nmtarr/gap_species/blob/master/DraftRangeSBMetadata.csv) in this repository but I think it's being held up by another pull request. 

Let me know what needs to be changed.  Here are some specific things to note.
- We need an abstract for the range data.
- Is the citation sufficient as is?
- I populated the range reviewer with the model reviewer's name.  That's generally how it worked.
- What should the IPDS be?
- In the range database, editor field, the names are currently very messy and inconsistent.  Hopefully, we can clean them up eventually.
- Did I get the doi in the range citation correct?

**Comment by skybristol** [comment]
I think it will be more efficient to generate the range items from the existing habitat map documents since those are now spruced up a bit following their initial creation. I have a script to do that, but we do need to resolve a few specific issues, clean some things up in the habitat map items, and then run the range items:

1) What is the "Global_SEQ_ID" in this new table?
2) We do need to decide on a simple, reasonable abstract that we can build from a pattern.
3) I'll work out an appropriate citation for the ranges using the pattern from the habitat maps, but we'll also need to go back and update the habitat map items to use the "Gap Analysis Project" convention.
4) We won't worry about the IPDS identifier for now.
5) In the habitat map items, we turned the "reviewer" type contact into "modeler." Do we want to call those same people reviewers in the ranges and modelers in the habitat maps, or is there one single type value that we want to use?
6) Now would be a good time to go ahead and clean up editors for these items. What we're looking for here are really firm contacts in the ScienceBase Directory that we can link to and keep up to date over time as needed. It does not look like that many distinct names in the CSV output, so we should be able to pick out the unique names, check contact information in the ScienceBase Directory, update that as needed, and then assemble the editor contacts for the specific range items from those records.

Here's an output from a script that cleans up and assembles the unique name strings into a list:

['Ken Boykin',
 '',
 'K Boykin',
 'Matthew Rubino',
 'Thomas Laxon',
 'Steven G Williams',
 'Thomas Laxson',
 'Nathan Tarr',
 'test',
 'Matt Rubino',
 'R. Adair',
 'Robert Adair',
 'Thomas',
 'Jeff Lonneker',
 'test32',
 'Rob Adair',
 'Jeff',
 'Robert Adair and Matt Rubino',
 'R. Adair and M. Rubino',
 'Test',
 'mACRAc',
 'Jocelyn Aycrigg',
 'Unknown',
 'K boykin',
 'K Boykin ',
 'kboykin',
 'Kboykin',
 'KBoykin']

There aren't that many strings to deal with here, and we can simply assemble a dictionary of the names (or even partial names since we would just be finding within a string) we want to look for with the appropriate ScienceBase Directory ID to use in cleaning the data into the ScienceBase Items. Once those are set, we should begin treating the ScienceBase collection items as the master source, writing any necessary data back out to backend databases for other uses.

**Comment by nmtarr** [comment]
Am I hearing that you don't need a table like the one I created in order to build the range items?  If so, no problem at all.  If you do need a table, what fields exactly do you want?  

1. Global_SEQ_ID is one of a few NatureServe codes in our databases.  I'll be honest, I don't know which NatureServe code is of interest here or for the TIR.  I know @SGWilliams is working on updating some of the taxonomy-related codes, but I'm not sure which one.  I took a guess at which ones were important and I guess I was wrong!  It will be nice when we can clean all those trash columns out of our databases.

4. I've removed the IPDS column

5. I think there were some cases where the person who edited a range was different than than who built the model.  In the databases, the model editor (modeler), model reviewer, and range editor are independent.  There is no field for range reviewer.  In the range csv I shared, the "editor" was populated from the range editor field.  I populated the "reviewer" field with the name of the person who reviewed the _model_.  I think this is appropriate because when we reviewed someone's model, we often had comments about the range.  Does that clear it up?  If it would be clearer to use words other than "editor" and "reviewer" here then that would be fine.

6.  Making a dictionary to handle those name errors/inconsistencies is a good idea.  I'll work that into my script.  I should have been able to fix them fairly quickly in the database but there are some unknown (to me) relationships between tables that keep me from making changes.  We can try again to fix these errors when we get around to overhauling our back end databases and in the future, we'll have ways to store, look up, and restrict names entered into such fields.

**Comment by skybristol** [comment]
We do need a table with any information that is specific to the ranges as opposed to the habitat maps. I'm doing some cleanup on the habitat map items in the [master collection](https://www.sciencebase.gov/catalog/item/527d0a83e4b0850ea0518326) that we will carry over to the range items.

1) I think I've got the situation on NatureServe IDs nailed down now after seeing Steve's explanation and cleaned up table with GAP IDs to NatureServe and ITIS IDs. I'll be feeding those back into the ScienceBase Items here shortly.

3) It sounds like the editor column from your table has the information we need that is unique to the range items. So, once we do a little bit of work to provide contact information for the unique editors and match that to the variable strings, we can work that into the ScienceBase information. It sounds like we should end up with a "reviewer" type contact and one or more "editor" type contacts for some of the ranges where your database has captured that information. The "reviewer" for a range item happens to be the same person as the "modeler" on a habitat map item, but this designation will make responsibility clear in the metadata.

4) The important thing at this point is to get the information on the ScienceBase Items as correct as possible for review and release. For the purposes of getting this stuff into the NCSU or BSU backends, I'd recommend simply pulling whatever metadata you need associated with the cleaned up species records in ScienceBase and using it from there.

**Comment by SGWilliams** [comment]
I'm on the verge of updating our SQL tables, some of those fields will be in the HabMap and Range SB item CSVs. Listed here with a brief explanation:

strScientificName (GAP SciName)
strSubSciNameText (GAP SubSciName)
strCommonName (GAP ComName)
intELEMENT_GLOBAL_SEQ_UID (NS Global ID)
strELCODE_BCD (NS Element Code)
strITIScode (TSN)
cbxNSGvalidMatch (checkbox indicating a correct matching of species' concept between GAP and NS)
cbxITISvalidMatch (checkbox indicating a correct matching of species' concept between GAP and ITIS)

Regarding Editor/Reviewer fields: I think having the same reviewer on both Range and HabMap items makes sense as Nathan stated. We simply need to drop the label of "Modeler" from the Range items and refer to the fields as "Range Editor", "Range Reviewer", "Model Editor" and "Model Reviewer" (with both "Range Reviewer" and "Model Reviewer" coming from the same field in the dB).

**Comment by SGWilliams** [comment]
We now have updated SB item csv files for both HabMaps and Ranges in the code repository pending a pull request I have going. Updates have happened to the following fields in bold italics:

### ScienceBaseHabMapCSV_20171024.csv
GAP_code
_**common_name**_
_**scientific_name**_
start_date
end_date
publication_date
_**citation**_
place_keywords
_**theme_keywords**_
_**editor**_
_**reviewer**_
_**NatureServe_element_code**_
_**TSN_code**_
_**Global_SEQ_ID**_
input_data
IPDS
_**abstract**_
_**doi**_
ScienceBase_url

### RangeSBMetadata23Oct2017.csv
GAP_code
_**reviewer**_
_**editors**_

We tried to begin updating the SB fields but stopped since we're running into items that were not initially defined (TSN_code, Global_SEQ_ID) so we stopped as that gets beyond our experience.

We also noticed that the DOI links are no longer valid even though we had them working about two months ago when we corrected the format as per Julie's instruction. Not sure what's going on there - were they disabled?

**Comment by McKerrow** [comment]
It is my understanding the DOIs are on hold until the records are reviewed and ready for release.

**Comment by jpmagee** [comment]
Yes all DOIs are on hold pending completion of the data review.  We can't register (turn on) the DOIs until data is approved for release.

So, while the DOIs won't resolve to an actual "final" webpage where the user could access the data they will resolve to a page like the picture below.  This is if they are in the correct format (https://doi.org/10.5066/F7SX6BGR). If they are in the format currently showing up in the SB items (https://doi:10.5066/F70863ND) they won't return any result.  I am unsure if the DOI format showing up in the SB item comes from the csv or from the code? But it seems if the csv uses the correct format then you guys have done what you can.
![doi webpage](https://user-images.githubusercontent.com/31905270/31965034-447de0c0-b8c3-11e7-895e-dd94d653fc1d.JPG)

**Comment by skybristol** [comment]
In ScienceBase, the DOIs are listed in the items as identifiers, not as links. So, the identifier text string doi:10.5066/F70863ND is a perfectly valid DOI identifier that has been minted as "in process" and will be made live in the DOI registry once the associated item is released. To turn that into a link, you can use a number of HTTP URL root paths, including http://doi.org/ and http://dx.doi.org/ followed by the DOI itself (e.g., 10.5066/F70863ND).

**Comment by jpmagee** [comment]
Just wanted to make clear that I was referring to the link to a DOI that is found in the citation section of the SB items. Not to the actual identifier in the SB item. Perhaps we want to take that link out? In the case of the DOI links in the citation section they are showing up like this https://doi:10.5066/F7MP51PC  If a user tries to use that as a valid link it won't take them anywhere. So do we want a DOI link in the citation section or just include it as an identifier? If we keep it in the citation we should consider the form in case someone tries to use that. 

Example of DOI link: U.S. Geological Survey - Gap Analysis Project, 2017, Alabama Beach Deermouse (Peromyscus polionotus ammobates) Habitat Map, U.S. Geological Survey data release, https://doi:10.5066/F7MP51PC.

**Comment by skybristol** [comment]
Oh yeah. We better fix that on the citations. I'll take a pass at those the next time I run code to update the habitat map items.

**Comment by skybristol** [comment]
@SGWilliams, I'm trying to get back to this in order to finish out the Habitat Map items and build the Range items. I'm looking at the latest CSV (https://github.com/usgs-bcb/gap_species/blob/master/ScienceBaseHabMapCSV_20171024.csv). I'm also looking at the JSON documents that @nmtarr generated and attached to the Habitat Map items. A couple things:

* There's no real need to include citation, abstract, place keywords, theme keywords, IPDS number, DOI, and publication date in the CSV. From my perspective, we let ScienceBase be the master source for that information. Those parts of the information are generated via code in ScienceBase from the foundational information output from the backend GAP database that you are giving us.
* It also seems like we should now rely on the process @nmtarr is working up to give us input data, assuming that is a more direct link (DOIs from ScienceBase items documenting and housing the source data) to what is in input_data in the spreadsheet.
* I might have missed something about the values in intGapNSmatch and intGapITISmatch in the GAP_ITIS_NS_Codes.csv. You now have values from 0-6 in there. What do those values signify? I think you said something about 0 and >0 somewhere, but can you provide definition on what 0-6 mean?

I will also have some additional questions for @nmtarr on some of the attributes in the species JSON documents that I either don't remember seeing before or just didn't pay enough attention to at the time.

**Comment by SGWilliams** [comment]
It was my understanding the items in the CSV were initially developed by Daniel to populate the SB items. The new version is just updating several of the fields and I didn't think there would be an issue to having all the original fields in there as the CSV is simply used to provide content to the upload (now update) script. Can't the update script just pull the fields it needs to update? I can remove those fields if need be - I may be missing the intent of the CSV. Just let me know.

I'm also a little confused on terms - what file(s) do you mean when you say "foundational information output from the backend GAP database"? I thought that was the data in the initial CSV. It certainly came from the backend dB.

I do see what you're saying about redundant input data information given the new JSON model records.

I initially planned to get the match codes and definitions into the CSV but we decided against it because we really weren't sure that's where it should go and it would have taken more time to implement. You can see the definitions in the issue "Work up code for generating GAP species range items #4"

My unending apologies for spraying information across multiple issues/repositories. But I am learning - at least I can recognize that now!

**Comment by skybristol** [comment]
It's fine having extra fields in the CSV. I didn't mean to confuse the issue. I'm just trying to get everything straight on the information we are using to hopefully not miss anything. Thanks for pointing to the writeup on the code definitions. I completely missed that.

I'm closing down this particular issue as I'm at least getting a bit confused. It's fantastic that we have this information all getting written down somewhere, though.
